printing_auto_base/models/printing_auto.py

```diff
@@ -0,0 +1,132 @@
+from odoo.osv import expression
+from odoo.tools.safe_eval import safe_eval
+
+
+class PrintingAuto(models.Model):
+    _name = "printing.auto"
+    _description = "Printing Auto"
+
+    name = fields.Char(string="Name", required=True)
+    file_type = fields.Selection(
+        [
+            ("report", "Report"),
+            ("attachment", "Attachment"),
+        ],
+        string="Type",
+        default="report",
+        required=True,
+        help=(
+            "Choose to print the result of an odoo report or a pre-existing "
+            "attachment (useful for labels received from carriers that are "
+            "recorded on the picking as an attachment)"
+        ),
+    )
+    field_object = fields.Char(
+        "Object", help="Select on which document the report must be executed"
+    )
+
+    report_id = fields.Many2one("ir.actions.report")
+
+    attachment_domain = fields.Char("Attachment", default="[]")
+    condition = fields.Char(
+        "Condition",
+        default="[]",
+        help="Give a domain that must be valid for printing this",
+    )
+
+    printer_id = fields.Many2one("printing.printer", string="Printer")
+    printer_tray_id = fields.Many2one("printing.tray")
+    nbr_of_copies = fields.Integer("Number of Copies", default=1)
+
+    label = fields.Boolean(string="Is Label")
+
+    @api.constrains("report_id", "file_type")
+    def check_report(self):
+        for rec in self:
+            if rec.file_type == "report" and not rec.report_id:
+                raise UserError(_("Report was not set"))
+
+    def do_print(self, record):
+        self.ensure_one()
+        record.ensure_one()
+
+        behaviour = self.get_behaviour()
+        printer = behaviour["printer"]
+        if not printer:
+            raise UserError(
+                _("No printer configured to print this {}.").format(self.name)
+            )
+
+        if self.nbr_of_copies <= 0:
+            return (printer, 0)
+        if not self._check_condition(record):
+            return (printer, 0)
+
+        count = 0
+        record = self._get_record(record)
+        for content in self._get_content(record):
+            for _n in range(self.nbr_of_copies):
+                printer.print_document(report=None, content=content, **behaviour)
+                count += 1
+        return (printer, count)
+
+    def get_behaviour(self):
+        if self.printer_id:
+            tray = self.printer_tray_id and self.printer_tray_id.system_name
+            return {"printer": self.printer_id, "tray": tray}
+        if self.file_type == "report":
+            return self.report_id.behaviour()
+        if self.label:
+            return {"printer": self.env.user.default_label_printer_id}
+        return self.env["ir.actions.report"]._get_user_default_print_behaviour()
+
+    def _get_record(self, record):
+        if self.field_object:
+            try:
+                return safe_eval(f"obj.{self.field_object}", {"obj": record})
```

Comment:
can't you simply use `mapped` and allow to pass a path to the field?

Reply:
not always that simple